### DISCUSSION TITLE
custom rules docs

### DISCUSSION THREAD

**Discussion by stropho** [monorepolint/monorepolint#201] [Q&A] [answered]
Hello,
thanks for the tool, it looks promising although I haven't tried yet. 

Could anybody give us some basic instructions how to write and use custom rules?
In case we make it work, we could probably create a PR with the docs 😉

**Comment by ericanderson** [comment, marked as answer]
Hello.

I fully intended to write some docs for you tonight but I realized all the docs committed reflected things from the 0.4 series while `main` is tracking the 0.5.0 series (which you can get on NPM btw). And I decided to fix that first. See #200 

The rules built in are a decent starting place to investigate the future. There is some cruft in there to support back compat but for a new rule you basically just need to export a class.

An easy example would be https://github.com/monorepolint/monorepolint/blob/main/packages/rules/src/alphabeticalDependencies.ts

You would need to implement `NewRuleModule` from `@monorepolint/config`: https://github.com/monorepolint/monorepolint/blob/main/packages/config/src/Config.ts#L33-L40

The check function does all the work. The most important thing in there is to use the context for everything, including file io. On a large monorepo it can get very expensive to keep reading/writing json files and the context has cached methods on it under `host`.

Here is an example of a fixer on an error: https://github.com/monorepolint/monorepolint/blob/main/packages/rules/src/packageEntry.ts#L54-L65
Note that it is using `mutateJson` helper to let you operate on the json as an input and a return value.

HOpe this helps.
